docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -0,0 +1,260 @@
+
+#### As an operator, I want to be able to be more inline with security compliance as I spin up nodes
+
+As an operator I would like to be able to have my node’s OS image cache for the ability to avoid security alerts, cache security updates and patches so that I’ll be more security compliant and up to date when I have to spin up a new node. Otherwise, flags will be raised that a node is out of date on boot up before it finishes patching and a temporary security risk will exist.
+
+### Implementation Details/Notes/Constraints
+
+The plan is to modify the existing Controllers with the Node Prototype Pattern as desired. These controller additions can be added to AzureMachinePool Controller.
+
+An operator will be able to decide to turn the feature on or off with an environment variable on clusterctl initialization, and then on a cluster by cluster basis can alter the AzureMachinePools to specify the use of the feature. They can also update the AzureMachinePool to customize how long they want the caching interval to be (see the yaml files below in this section for the caching interval).
+
+Example of the environment variable being turned on:
+
+```
+export AZURE_OS_CACHING=true
+```
+
+The controller will maintain a timestamp in each AzureMachinePool, and when the current time is the chosen interval ahead or more, the controller will perform the caching. Since the current controller manager requeues all objects every ten minutes by default the objects will be requeued shortly after its due time to be recached. This is because typically we expect to cache every 24 hours and it is very unexpected that this won't be frequent enough considering normal patch rates. 
+
+Example of how the timestamp will be maintained in the AzureMachinePool:
+
+```yaml
+status:
+  lastPrototype: "2023-06-12T23:14:55Z"
+```
+
+When the process is started it should go through the nodes of the cluster, choose a healthy node, shut it down, take a snapshot of it, restart it, create a compute image gallery image, delete the snapshot, and then configure the AzureMachinePool specs to use that compute image gallery image. After, it will store the current time as its timestamp. As a note, for the first implementation of this feature we will require the user to also use a compute image gallery image.
+
+Diagram of the Node OS Caching Process:
+
+![Figure 1](./images/node-os-image-cache.png)
+
+As for why the healthy node has to be shut down while creating a snapshot of it, if it isn’t shut down first then pods can be scheduled as the snapshot is taken which will cause some dangerous states in terms of how it exists after being utilized by the AzureMachinePools.
```

Comment:
All object names such as "AzureMachinePools" or variable references should ideally be formatted as code, i.e. `AzureMachinePools` , `metadata.creationTimestamp`, `latestModelApplied` , `true`

Reply:
This took a minute to format, but done!